samples/cpp/color_correction_model.cpp

```diff
@@ -0,0 +1,232 @@
+
+        // Read query image when using pre-computed CCM
+        queryImage = imread(findFile(queryFile));
+        if (queryImage.empty()) {
+            cout << "Error: Unable to read query image." << endl;
+            return -1;
+        }
+    } else {
+        // Read input image for computing new CCM
+        Mat originalImage = imread(findFile(inputFile));
+        if (originalImage.empty()) {
+            cout << "Error: Unable to read input image." << endl;
+            return -1;
+        }
+
+        // Process first image to compute CCM
+        Mat image = originalImage.clone();
+        Mat src;
+
+        Ptr<CCheckerDetector> detector;
+        if (!modelPath.empty() && !configPath.empty()) {
+            Net net = readNetFromTensorflow(modelPath, configPath);
+            net.setPreferableBackend(getBackendID(backend));
+            net.setPreferableTarget(getTargetID(target));
+            detector = CCheckerDetector::create(net);
+            cout << "Using DNN-based checker detector." << endl;
+        } else {
+            detector = CCheckerDetector::create();
+            cout << "Using thresholding-based checker detector." << endl;
+        }
+        detector->setColorChartType(chartType);
+
+        if (!processFrame(image, detector, src, nc)) {
+            cout << "No chart detected in the input image!" << endl;
+            return -1;
+        }
+        // Convert to double and normalize
+        src.convertTo(src, CV_64F, 1.0/255.0);
+
+        // Color correction model
+        model = ColorCorrectionModel(src, COLORCHECKER_MACBETH);
+        model.setCcmType(CCM_LINEAR);
+        model.setDistance(DISTANCE_CIE2000);
+        model.setLinearization(LINEARIZATION_GAMMA);
+        model.setLinearizationGamma(2.2);
+
+        Mat ccm = model.compute();
+        cout << "Computed CCM Matrix:\n" << ccm << endl;
+        cout << "Loss: " << model.getLoss() << endl;
+
+        // Save model parameters to YAML file
+        FileStorage fs("ccm_output.yaml", FileStorage::WRITE);
+        model.write(fs);
+        fs.release();
+        cout << "Model parameters saved to ccm_output.yaml" << endl;
+
+        // Set query image for correction
+        if (queryFile.empty()) {
+            cout << "[WARN] No query image provided, applying color correction on input image" << endl;
+            queryImage = originalImage.clone();
+        } else {
+            queryImage = imread(findFile(queryFile));
+            if (queryImage.empty()) {
+                cout << "Error: Unable to read query image." << endl;
+                return -1;
+            }
+        }
+    }
+
+    // Apply correction to query image
+    Mat calibratedImage, normalizedImage;
+    cvtColor(queryImage, normalizedImage, COLOR_BGR2RGB);
+    normalizedImage.convertTo(normalizedImage, CV_64F, 1.0/255.0);  // Convert to double and normalize
+    model.correctImage(normalizedImage, calibratedImage);
```

Comment:
I propose to Move BRG/RGB logic into model with a flag (default to bgr). All OpenCV functions are designed for BGR. The flag allows to make optimizations without API change.